apps/prairielearn/src/lib/question-render.ts

```diff
@@ -396,66 +395,93 @@ function buildLocals({
+  const resultLocals = locals as any;
+
+  const question_course = await getQuestionCourse(locals.question, locals.course);
+  resultLocals.question_is_shared = await sqldb.queryRow(
     sql.select_is_shared,
     { question_id: locals.question.id },
     z.boolean(),
   );
 
-  if (variant_id != null) {
-    locals.variant = await sqldb.queryOptionalRow(
-      sql.select_variant_for_render,
-      {
-        variant_id,
-        question_id: locals.question.id,
-        instance_question_id: locals.instance_question?.id,
-      },
-      VariantSelectResultSchema,
-    );
-    if (locals.variant == null) {
-      throw new error.HttpStatusError(404, 'Variant not found');
+  const variant = await run(async () => {
+    if (variant_id != null) {
+      return await sqldb.queryOptionalRow(
+        sql.select_variant_for_render,
+        {
+          variant_id,
+          question_id: locals.question.id,
+          instance_question_id: locals.instance_question?.id,
+        },
+        VariantSelectResultSchema,
+      );
+    } else {
+      const require_open = !!locals.assessment && locals.assessment.type !== 'Exam';
+      const instance_question_id = locals.instance_question?.id ?? null;
+      const course_instance_id = locals.course_instance_id ?? locals.course_instance?.id ?? null;
+      const options = { variant_seed };
+      return await ensureVariant(
+        locals.question.id,
+        instance_question_id,
+        locals.user.user_id,
+        locals.authn_user.user_id,
+        course_instance_id,
+        locals.course,
+        question_course,
+        options,
+        require_open,
+        locals.client_fingerprint_id ?? null,
+      );
     }
-  } else {
-    const require_open = locals.assessment && locals.assessment.type !== 'Exam';
-    const instance_question_id = locals.instance_question?.id;
-    const course_instance_id = locals.course_instance_id ?? locals.course_instance?.id ?? null;
-    const options = { variant_seed };
-    locals.variant = await ensureVariant(
-      locals.question.id,
-      instance_question_id,
-      locals.user.user_id,
-      locals.authn_user.user_id,
-      course_instance_id,
-      locals.course,
-      locals.question_course,
-      options,
-      require_open,
-      locals.client_fingerprint_id,
-    );
-  }
+  });
+
+  if (variant == null) throw new error.HttpStatusError(404, 'Variant not found');
+
+  resultLocals.variant = variant;
 
   const {
     urlPrefix,
-    variant,
+    course,
+    course_instance,
     question,
     instance_question,
     assessment,
     assessment_instance,
     assessment_question,
     group_config,
     group_role_permissions,
+    authz_data,
     authz_result,
   } = locals;
 
-  const urls = buildQuestionUrls(urlPrefix, variant, question, instance_question);
+  const urls = buildQuestionUrls(urlPrefix, variant, question, instance_question ?? null);
+  Object.assign(urls, options?.urlOverrides);
   Object.assign(locals, urls);
```

Comment:
Any reason why you switched the order? Shouldn't the overrides be assigned after the built URLs?

Reply:
Please disregard, this is correct.